sw/host/rom_ext_image_tools/signer/src/main.rs

```diff
@@ -9,91 +9,188 @@
-    let system_state_value =
-        &system_state_value(&config.input_files.system_state_value_path);
+struct Args {
+    input: PathBuf,
+    priv_key: PathBuf,
+    output: PathBuf,
+}
 
-    let mut message_to_sign = Vec::<u8>::new();
-    message_to_sign.extend_from_slice(system_state_value);
-    message_to_sign.extend_from_slice(device_usage_value);
-    message_to_sign.extend_from_slice(image_sign_data);
+impl Args {
+    pub fn new(args: env::Args) -> Result<Args> {
+        let args = args.skip(1).collect::<Vec<_>>();
+        match args.as_slice() {
+            [input, priv_key, output] => Ok(Args {
+                    input: PathBuf::from(input),
+                    priv_key: PathBuf::from(priv_key),
+                    output: PathBuf::from(output),
+                }),
+            args => bail!("Expected exactly 3 positional arguments: input, private key, and output, got: {}.", args.len()),
+        }
+    }
+}
 
-    let signature = RsaSignature::<B3072, RsaPkcs1v15, Sha256>::sign(
-        &private_key,
-        &message_to_sign,
-    )
-    .expect("Failed to sign!");
+/// Parses an unsigned big-endian hex string into a little-endian byte vector.
+fn parse_hex_str(hex_str: &str) -> Result<Vec<u8>> {
+    ensure!(
+        hex_str.starts_with("0x") && hex_str.len() >= 3,
+        "Invalid hex string: {}",
+        hex_str
+    );
+
+    let mut bytes = Vec::new();
+    hex_str
+        .as_bytes()
+        .iter()
+        .skip(2)
+        .rev()
+        .map(|digit| match digit {
+            b'0'..=b'9' => Ok(digit - b'0'),
+            b'a'..=b'f' => Ok(digit - b'a' + 10),
+            b'A'..=b'F' => Ok(digit - b'A' + 10),
+            _ => bail!("Invalid hex digit: {} ({}).", *digit as char, digit),
+        })
+        .enumerate()
+        .try_for_each(|(i, digit)| -> Result<()> {
+            match i % 2 {
+                0 => bytes.push(digit?),
+                _ => *bytes.last_mut().unwrap() |= digit? << 4,
+            };
+            Ok(())
+        })?;
+    Ok(bytes)
+}
 
-    image.update_signature_field(&signature.bytes());
+/// Updates the manifest of an image.
+// TODO: This function must use a public key.
+fn update_image_manifest(image: &mut Image, key: impl AsRef<Path>) -> Result<()> {
+    let key = fs::read(&key).with_context(|| {
+        format!("Failed to read the key from `{}`.", key.as_ref().display())
+    })?;
+    let key =
+        PrivateKey::parse_from_der(&key).context("Failed to parse the key.")?;
+
+    image.set_manifest_field(
+        &manifest::ROM_EXT_SIGNATURE_KEY_PUBLIC_EXPONENT,
+        key.public_exponent_be().iter().rev().cloned(),
+    )?;
+    image.set_manifest_field(
+        &manifest::ROM_EXT_SIGNATURE_KEY_MODULUS,
+        key.public_modulus_be().iter().rev().cloned(),
+    )?;
+    image.set_manifest_field(
+        &manifest::ROM_EXT_MANIFEST_IDENTIFIER,
+        parse_hex_str("0x4552544f")?,
```

Comment:
I guess these are temporary hardcoded values?

Reply:
For ROM_EXT, this value should always be the same ('OTRE'), however, we're planning to use the same manifest stucture for identifying and validating owner BL0.  We'll probably want to pick a different ID for boot-loaders.

Reply:
@cfrantz, yes. Fields that are expected to be passed in use `parse_hex_str` (with two exceptions, I think, because they were longer than other fields).